AmmadAmjad-CodeTest/Server/controllers/middleware.js

```diff
@@ -0,0 +1,72 @@
+var cryptojs = require('crypto-js');
+
+module.exports = function(db) {
+  return {
+    logger: function(req, res, next) {
+      console.log(
+        "Request : " +
+        new Date().toString() +
+        " " +
+        req.method +
+        " " +
+        req.originalUrl
+      );
+      next();
+    },
+    optionalAuthentication: function(req , res , next){
+      var token = req.get("Authorization") || "";
+      if (token === undefined || token === "") {
+        next();
+        return;
+      }
+      db.user
+        .findOne({
+          where: {
+            tokenHash: cryptojs.MD5(token).toString()
+          }
+        })
+        .then(function(user) {
+          if (!user) {
+            next();
+            return;
+          }
+          req.user = user;
+          next();
+        })
+        .catch(function() {
+          next();
+          return;
+        });
+    },
+    requireAuthentication: function(req, res, next) {
+      var token = req.get("Authorization") || "";
+      if (token === undefined || token === "") {
+        res.status(401).send({
+          message: "Please include user token in header as Authentication"
+        });
+        return;
+      }
+      db.user
+        .findOne({
+          where: {
+            tokenHash: cryptojs.MD5(token).toString()
```

Comment:
Since we are storing token in database on user's `tokenHash` field, does this mean we don't want the user to have multiple sessions (multiple browsers / devices)?